lib/vfscore/README.md

```diff
@@ -0,0 +1,59 @@
+# VFSCore: Unikraft's Virtual File System API
+
+The `vfscore` is an internal library of Unikraft that provides a generic interface of system calls related to filesystem management.
+There are multiple types of filesystems available in Unikraft (e.g: `ramfs`,` 9pfs`).
+
+## Key Functions and Data Structures
+
+The `vfscore` library provides 2 structures to define operations on a filesystem (defined in `lib/vfscore/include/vfscore/mount.h`):
+
+```c
+struct vfsops {
+        int (*vfs_mount)        (struct mount *, const char *, int, const void*);
+        ...
+        struct vnops    *vfs_vnops;
+};
+```
+
+```c
+struct vnops {
+        vnop_open_t             vop_open;
+        vnop_close_t            vop_close;
+        vnop_read_t             vop_read;
+        vnop_write_t            vop_write;
+        vnop_seek_t             vop_seek;
+        vnop_ioctl_t            vop_ioctl;
+        ...
+};
+```
+
+The first structure mainly defines the operation of mounting the filesystem, while the second defines the operations that can be executed on files (regular files, directories, etc).
+The `vnops` structure can be seen as the `file_operation` structure in the Linux Kernel (more as an idea).
+More about this structure [here](https://tldp.org/LDP/lkmpg/2.4/html/c577.htm).
+
+The filesystem library will define two such structures through which it will provide the specified operations.
+To understand how these operations end up being used let's examine the open system call:
+
+```c
+int
+sys_open(char *path, int flags, mode_t mode, struct vfscore_file **fpp)
+{
+        struct vfscore_file *fp;
+        struct vnode *vp;
+        ...
+        error = VOP_OPEN(vp, fp);
+}
+```
+
+`VOP_OPEN()` is a macro that is defined as follows:
+
+```c
+#define VOP_OPEN(VP, FP)           ((VP)->v_op->vop_open)(FP)
+```
```

Comment:
I would add an example of how the `VOP_OPEN` is defined in an actual filesystem (e.g. ramfs).
Just explain one of the declarations [here](https://github.com/unikraft/unikraft/blob/staging/lib/ramfs/ramfs_vnops.c#L638-L676), on how the `ramfs_vnops` structure is defined. Same for [`vfsops`](https://github.com/unikraft/unikraft/blob/staging/lib/ramfs/ramfs_vfsops.c#L49-L63).
I would also add a few lines on how to register a new filesystem (see [here](https://github.com/unikraft/unikraft/blob/staging/lib/ramfs/ramfs_vfsops.c#L65-L71) for `ramfs`).